File: ios/ExpoCallKitModule.swift
import ExpoModulesCore
import Foundation

// MARK: - Records (JS → native argument shapes)

struct ParticipantRecord: Record {
  @Field
  var id: String = ""

  @Field
  var displayName: String?

  @Field
  var phoneNumber: String?

  @Field
  var email: String?

  func asParticipant() -> Participant {
    Participant(
      id: id,
      displayName: displayName,
      phoneNumber: phoneNumber,
      email: email
    )
  }
}

struct IncomingCallRecord: Record {
  @Field
  var eventId: String = ""

  @Field
  var serverCallId: String = ""

  @Field
  var caller: ParticipantRecord = ParticipantRecord()

  @Field
  var hasVideo: Bool = false

  @Field
  var metadata: [String: Any]?

  func asRingPayload() throws -> RingPayload {
    guard !eventId.isEmpty, !serverCallId.isEmpty, !caller.id.isEmpty else {
      throw InvalidPayloadException()
    }
    return RingPayload(
      eventId: eventId,
      serverCallId: serverCallId,
      caller: caller.asParticipant(),
      hasVideo: hasVideo,
      metadata: metadata
    )
  }
}

struct OutgoingCallOptionsRecord: Record {
  @Field
  var hasVideo: Bool = false

  @Field
  var metadata: [String: Any]?
}

// MARK: - Module

public final class ExpoCallKitModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoCallKit")

    Events(
      CKEvent.incomingCall,
      CKEvent.callAnswered,
      CKEvent.callEnded,
      CKEvent.outgoingCallStarted,
      CKEvent.muteChanged,
      CKEvent.holdChanged,
      CKEvent.dtmf,
      CKEvent.audioSessionActivated,
      CKEvent.audioSessionDeactivated,
      CKEvent.voipTokenUpdated
    )

    OnCreate {
      EventHub.shared.attach(self)
    }

    OnDestroy {
      EventHub.shared.attach(nil)
    }

    // Per-event observation drives the replay buffers in EventHub: each
    // event flushes independently the moment JS mounts its listener.
    OnStartObserving(CKEvent.incomingCall) {
      EventHub.shared.startObserving(CKEvent.incomingCall)
    }
    OnStopObserving(CKEvent.incomingCall) {
      EventHub.shared.stopObserving(CKEvent.incomingCall)
    }

    OnStartObserving(CKEvent.callAnswered) {
      EventHub.shared.startObserving(CKEvent.callAnswered)
    }
    OnStopObserving(CKEvent.callAnswered) {
      EventHub.shared.stopObserving(CKEvent.callAnswered)
    }

    OnStartObserving(CKEvent.callEnded) {
      EventHub.shared.startObserving(CKEvent.callEnded)
    }
    OnStopObserving(CKEvent.callEnded) {
      EventHub.shared.stopObserving(CKEvent.callEnded)
    }

    OnStartObserving(CKEvent.outgoingCallStarted) {
      EventHub.shared.startObserving(CKEvent.outgoingCallStarted)
    }
    OnStopObserving(CKEvent.outgoingCallStarted) {
      EventHub.shared.stopObserving(CKEvent.outgoingCallStarted)
    }

    OnStartObserving(CKEvent.muteChanged) {
      EventHub.shared.startObserving(CKEvent.muteChanged)
    }
    OnStopObserving(CKEvent.muteChanged) {
      EventHub.shared.stopObserving(CKEvent.muteChanged)
    }

    OnStartObserving(CKEvent.holdChanged) {
      EventHub.shared.startObserving(CKEvent.holdChanged)
    }
    OnStopObserving(CKEvent.holdChanged) {
      EventHub.shared.stopObserving(CKEvent.holdChanged)
    }

    OnStartObserving(CKEvent.dtmf) {
      EventHub.shared.startObserving(CKEvent.dtmf)
    }
    OnStopObserving(CKEvent.dtmf) {
      EventHub.shared.stopObserving(CKEvent.dtmf)
    }

    OnStartObserving(CKEvent.audioSessionActivated) {
      EventHub.shared.startObserving(CKEvent.audioSessionActivated)
    }
    OnStopObserving(CKEvent.audioSessionActivated) {
      EventHub.shared.stopObserving(CKEvent.audioSessionActivated)
    }

    OnStartObserving(CKEvent.audioSessionDeactivated) {
      EventHub.shared.startObserving(CKEvent.audioSessionDeactivated)
    }
    OnStopObserving(CKEvent.audioSessionDeactivated) {
      EventHub.shared.stopObserving(CKEvent.audioSessionDeactivated)
    }

    OnStartObserving(CKEvent.voipTokenUpdated) {
      EventHub.shared.startObserving(CKEvent.voipTokenUpdated)
    }
    OnStopObserving(CKEvent.voipTokenUpdated) {
      EventHub.shared.stopObserving(CKEvent.voipTokenUpdated)
    }

    // MARK: Calls

    AsyncFunction("reportIncomingCall") { (record: IncomingCallRecord) -> String in
      let payload = try record.asRingPayload()
      let id = try await CallCenter.shared.reportIncomingCallFromJS(payload)
      return id.uuidString.lowercased()
    }

    AsyncFunction("startOutgoingCall") {
      (recipient: ParticipantRecord, options: OutgoingCallOptionsRecord?) -> String in
      let id = try await CallCenter.shared.startOutgoingCall(
        recipient: recipient.asParticipant(),
        hasVideo: options?.hasVideo ?? false,
        metadata: options?.metadata
      )
      return id.uuidString.lowercased()
    }

    AsyncFunction("reportOutgoingCallConnected") { (callId: String) in
      try CallCenter.shared.reportOutgoingConnected(try Self.parseUuid(callId))
    }

    AsyncFunction("answerAcknowledged") { (requestId: String) in
      // No-op when the request already resolved (raced a timeout).
      _ = await PendingAnswers.shared.acknowledge(try Self.parseUuid(requestId))
    }

    AsyncFunction("answerFailed") { (requestId: String) in
      _ = await PendingAnswers.shared.fail(try Self.parseUuid(requestId))
    }

    AsyncFunction("endCall") { (callId: String) in
      try await CallCenter.shared.endCall(try Self.parseUuid(callId))
    }

    AsyncFunction("reportCallEnded") { (callId: String, reason: String) in
      guard let endReason = EndReason(rawValue: reason) else {
        throw InvalidEndReasonException(reason)
      }
      try CallCenter.shared.reportCallEnded(try Self.parseUuid(callId), reason: endReason)
    }

    AsyncFunction("setMuted") { (callId: String, muted: Bool) in
      try await CallCenter.shared.setMuted(try Self.parseUuid(callId), muted: muted)
    }

    AsyncFunction("setOnHold") { (callId: String, onHold: Bool) in
      try await CallCenter.shared.setOnHold(try Self.parseUuid(callId), onHold: onHold)
    }

    AsyncFunction("getActiveCall") { () -> [String: Any]? in
      CallCenter.shared.activeCallSession()
    }

    // MARK: Push / audio / permissions

    Function("getVoipToken") { () -> [String: Any]? in
      guard let token = VoipPushCoordinator.shared.currentToken else {
        return nil
      }
      return ["token": token, "type": "apns-voip"]
    }

    Function("registerVoipPushes") {
      VoipPushCoordinator.shared.register()
    }

    Function("configureAudioSession") {
      AudioSessionCoordinator.shared.prewarm()
    }

    AsyncFunction("requestPermissions") { () -> [String: String] in
      // Nothing to request on iOS: CallKit UI needs no runtime permission and
      // the mic prompt is triggered by the app's media layer on first use.
      ["notifications": "granted"]
    }
  }

  private static func parseUuid(_ value: String) throws -> UUID {
    guard let uuid = UUID(uuidString: value) else {
      throw InvalidUuidException(value)
    }
    return uuid
  }
}

File: android/src/main/java/dev/rossslaney/expocallkit/ExpoCallKitModule.kt
package dev.rossslaney.expocallkit

import android.Manifest
import android.content.Intent
import android.os.Build
import expo.modules.interfaces.permissions.PermissionsStatus
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID

class ExpoCallKitModule : Module() {
    override fun definition() = ModuleDefinition {
        Name("ExpoCallKit")

        Events(
            CKEvents.INCOMING_CALL,
            CKEvents.CALL_ANSWERED,
            CKEvents.CALL_ENDED,
            CKEvents.OUTGOING_CALL_STARTED,
            CKEvents.MUTE_CHANGED,
            CKEvents.HOLD_CHANGED,
            CKEvents.DTMF,
            CKEvents.AUDIO_SESSION_ACTIVATED,
            CKEvents.AUDIO_SESSION_DEACTIVATED,
            CKEvents.AUDIO_ROUTE_CHANGED,
            CKEvents.VOIP_TOKEN_UPDATED,
        )

        OnCreate {
            appContext.reactContext?.let { CallEngine.initialize(it) }
            EventHub.attachSender { name, body -> sendEvent(name, body) }
            // Cold-start answer: the notification's answer action IS the
            // launch intent, so OnNewIntent never fires for it.
            consumeAnswerIntent(appContext.currentActivity?.intent)
        }

        OnDestroy {
            EventHub.attachSender(null)
        }

        OnNewIntent { intent ->
            consumeAnswerIntent(intent)
        }

        OnActivityEntersForeground {
            // Covers cold starts where the activity was not yet attached
            // during OnCreate. The action is cleared after processing, so
            // re-checks are harmless.
            consumeAnswerIntent(appContext.currentActivity?.intent)
        }

        // Per-event observation drives the replay buffers in EventHub.
        OnStartObserving(CKEvents.INCOMING_CALL) {
            EventHub.startObserving(CKEvents.INCOMING_CALL)
        }
        OnStopObserving(CKEvents.INCOMING_CALL) {
            EventHub.stopObserving(CKEvents.INCOMING_CALL)
        }

        OnStartObserving(CKEvents.CALL_ANSWERED) {
            EventHub.startObserving(CKEvents.CALL_ANSWERED)
        }
        OnStopObserving(CKEvents.CALL_ANSWERED) {
            EventHub.stopObserving(CKEvents.CALL_ANSWERED)
        }

        OnStartObserving(CKEvents.CALL_ENDED) {
            EventHub.startObserving(CKEvents.CALL_ENDED)
        }
        OnStopObserving(CKEvents.CALL_ENDED) {
            EventHub.stopObserving(CKEvents.CALL_ENDED)
        }

        OnStartObserving(CKEvents.OUTGOING_CALL_STARTED) {
            EventHub.startObserving(CKEvents.OUTGOING_CALL_STARTED)
        }
        OnStopObserving(CKEvents.OUTGOING_CALL_STARTED) {
            EventHub.stopObserving(CKEvents.OUTGOING_CALL_STARTED)
        }

        OnStartObserving(CKEvents.MUTE_CHANGED) {
            EventHub.startObserving(CKEvents.MUTE_CHANGED)
        }
        OnStopObserving(CKEvents.MUTE_CHANGED) {
            EventHub.stopObserving(CKEvents.MUTE_CHANGED)
        }

        OnStartObserving(CKEvents.HOLD_CHANGED) {
            EventHub.startObserving(CKEvents.HOLD_CHANGED)
        }
        OnStopObserving(CKEvents.HOLD_CHANGED) {
            EventHub.stopObserving(CKEvents.HOLD_CHANGED)
        }

        OnStartObserving(CKEvents.DTMF) {
            EventHub.startObserving(CKEvents.DTMF)
        }
        OnStopObserving(CKEvents.DTMF) {
            EventHub.stopObserving(CKEvents.DTMF)
        }

        OnStartObserving(CKEvents.AUDIO_SESSION_ACTIVATED) {
            EventHub.startObserving(CKEvents.AUDIO_SESSION_ACTIVATED)
        }
        OnStopObserving(CKEvents.AUDIO_SESSION_ACTIVATED) {
            EventHub.stopObserving(CKEvents.AUDIO_SESSION_ACTIVATED)
        }

        OnStartObserving(CKEvents.AUDIO_SESSION_DEACTIVATED) {
            EventHub.startObserving(CKEvents.AUDIO_SESSION_DEACTIVATED)
        }
        OnStopObserving(CKEvents.AUDIO_SESSION_DEACTIVATED) {
            EventHub.stopObserving(CKEvents.AUDIO_SESSION_DEACTIVATED)
        }

        OnStartObserving(CKEvents.AUDIO_ROUTE_CHANGED) {
            EventHub.startObserving(CKEvents.AUDIO_ROUTE_CHANGED)
        }
        OnStopObserving(CKEvents.AUDIO_ROUTE_CHANGED) {
            EventHub.stopObserving(CKEvents.AUDIO_ROUTE_CHANGED)
        }

        OnStartObserving(CKEvents.VOIP_TOKEN_UPDATED) {
            EventHub.startObserving(CKEvents.VOIP_TOKEN_UPDATED)
        }
        OnStopObserving(CKEvents.VOIP_TOKEN_UPDATED) {
            EventHub.stopObserving(CKEvents.VOIP_TOKEN_UPDATED)
        }

        // region Calls

        AsyncFunction("reportIncomingCall") { raw: Map<String, Any?> ->
            ensureInitialized()
            val payload = RingPayload.fromMap(raw) ?: throw InvalidPayloadError()
            CallEngine.reportIncomingCall(payload).toString()
        }

        AsyncFunction("startOutgoingCall") { recipient: Map<String, Any?>, options: Map<String, Any?>? ->
            ensureInitialized()
            val participant = Participant.fromMap(recipient) ?: throw InvalidPayloadError()
            @Suppress("UNCHECKED_CAST")
            CallEngine.startOutgoingCall(
                recipient = participant,
                hasVideo = options?.get("hasVideo") as? Boolean ?: false,
                metadata = options?.get("metadata") as? Map<String, Any?>,
            ).toString()
        }

        AsyncFunction("reportOutgoingCallConnected") { callId: String ->
            CallEngine.reportOutgoingConnected(parseUuid(callId))
        }

        AsyncFunction("answerAcknowledged") { requestId: String ->
            // No-op when the request already resolved (raced a timeout).
            CallEngine.acknowledgeAnswer(parseUuid(requestId))
        }

        AsyncFunction("answerFailed") { requestId: String ->
            CallEngine.rejectAnswer(parseUuid(requestId))
        }

        AsyncFunction("endCall") { callId: String ->
            CallEngine.endCall(parseUuid(callId))
        }

        AsyncFunction("reportCallEnded") { callId: String, reason: String ->
            val endReason = EndReason.fromWire(reason) ?: throw InvalidEndReasonError(reason)
            CallEngine.reportCallEnded(parseUuid(callId), endReason)
        }

        AsyncFunction("setMuted") { callId: String, muted: Boolean ->
            CallEngine.setMuted(parseUuid(callId), muted)
        }

        AsyncFunction("setOnHold") { callId: String, onHold: Boolean ->
            CallEngine.setOnHold(parseUuid(callId), onHold)
        }

        AsyncFunction("getActiveCall") {
            CallEngine.activeSessionMap()
        }

        // endregion

        // region Push / audio / permissions

        Function("getVoipToken") {
            // Push token plumbing is the app's responsibility on Android
            // (no FCM service ships in this module).
            null as Map<String, Any?>?
        }

        Function("registerVoipPushes") {
            // No-op on Android; see getVoipToken.
        }

        Function("configureAudioSession") {
            // No-op on Android; core-telecom owns audio focus.
        }

        AsyncFunction("getAudioRouteState") {
            CallEngine.audioRouteState()
        }

        AsyncFunction("selectAudioRoute") { callId: String, routeId: String, promise: Promise ->
            val parsedId = try {
                parseUuid(callId)
            } catch (error: CodedException) {
                promise.reject(error)
                return@AsyncFunction
            }

            CallEngine.selectAudioRoute(parsedId, routeId) { result ->
                result.fold(
                    onSuccess = { promise.resolve() },
                    onFailure = { error ->
                        promise.reject(
                            error as? CodedException
                                ?: AudioRouteRejectedError(error.message ?: "request failed"),
                        )
                    },
                )
            }
        }

        AsyncFunction("setSpeakerEnabled") { callId: String, enabled: Boolean ->
            CallEngine.setSpeakerEnabled(parseUuid(callId), enabled)
        }

        AsyncFunction("playCallFeedback") { callId: String ->
            CallEngine.playCallFeedback(parseUuid(callId))
        }

        AsyncFunction("requestPermissions") { promise: Promise ->
            if (Build.VERSION.SDK_INT < 33) {
                promise.resolve(mapOf("notifications" to "granted"))
                return@AsyncFunction
            }
            val permissions = appContext.permissions
            if (permissions == null) {
                promise.resolve(mapOf("notifications" to "undetermined"))
                return@AsyncFunction
            }
            permissions.askForPermissions(
                { results ->
                    val status = results[Manifest.permission.POST_NOTIFICATIONS]?.status
                    val outcome = if (status == PermissionsStatus.GRANTED) "granted" else "denied"
                    promise.resolve(mapOf("notifications" to outcome))
                },
                Manifest.permission.POST_NOTIFICATIONS,
            )
        }

        // endregion
    }

    private fun ensureInitialized() {
        appContext.reactContext?.let { CallEngine.initialize(it) }
    }

    /** Handles the notification answer action; clears it to avoid replays. */
    private fun consumeAnswerIntent(intent: Intent?) {
        if (intent?.action != CKIntents.ACTION_ANSWER) {
            return
        }
        val id = intent.getStringExtra(CKIntents.EXTRA_CALL_ID)?.let {
            try {
                UUID.fromString(it)
            } catch (_: IllegalArgumentException) {
                null
            }
        } ?: return

        intent.action = null
        CallEngine.handleAnswer(id)
    }

    private fun parseUuid(value: String): UUID = try {
        UUID.fromString(value)
    } catch (_: IllegalArgumentException) {
        throw InvalidUuidError(value)
    }
}
